### ISSUE TITLE
Unexpected literal type narrowing within generator expression

### ISSUE THREAD

**Issue by kltws** [microsoft/pyright#6109]
```python
from enum import Enum
from typing import Any, Iterable

class TestEnum(Enum):
    L = ...

def do_something_with_literals(iterable: Iterable[Any]):
    literals = (i for i in iterable if i is TestEnum.L)
    ...
```
`literals` is inferred as `Generator[Never, None, None]`.
Expected `Generator[Literal[TestEnum.L], None, None]` (ideally) or `Generator[Any, None, None]` (not as good, but better than Never), since the advanced type concepts page states that "x is E (where E is a literal enum or bool)" type narrowing is performed.

(Not my use case, but the same thing happens with literal bools)
```python
def do_something_with_literals(iterable: Iterable[Any]):
    literals = (i for i in iterable if i is True)   # Generator[Never, None, None]
```

(With Pylance v2023.10.11)

**Comment by erictraut** [comment]
Thanks for the bug report. This issue wasn't confined to generators. It was a more general problem with the `x is <Literal>` type guard logic, which wasn't properly handling the case where `x` was `Any`.

```python
def func(x: Any):
    if x is TestEnum.L:
        reveal_type(x)  # Should be Literal[TestEnum.L]
    else:
        reveal_type(x)  # Should be Any
```

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.331, which I just published. It will also be included in a future release of pylance.
